### PR TITLE
Dynamic modules - throw on star exports

### PR DESCRIPTION
**Pull request by guybedford**

### CODE REVIEW

**Comment by guybedford** [comment]
(the last commit is the new change)

File: spec.html [hunk 1]
```diff
@@ -8444,7 +8444,7 @@ <h1>Module Namespace Exotic Objects</h1>
               [[Exports]]
             </td>
             <td>
-              List of String
+              List of String | *null*
```
Comment (littledan) [inline spec.html:8447]: Doesn't the need for all this | null/extensibility stuff go away?

File: spec.html [hunk 2]
```diff
@@ -22034,6 +22048,7 @@ <h1>GetExportedNames ( _exportStarSet_ ) Concrete Method</h1>
           </emu-alg>
           <emu-note>
             <p>GetExportedNames does not filter out or throw an exception for names that have ambiguous star export bindings.</p>
+            <p>A *null* return from GetExportedNames indicates a non-Source-Text Module Record does not support star exports. This throws a corresponding SyntaxError exception.</p>
```
Comment (littledan) [inline spec.html:22051]: Nit: This might be clearer as a separate note

File: spec.html (hunk 1 above)
Comment (littledan) [reply spec.html:8447]: Ah, never mind, no, this is how CJS modules are directly (it is just never that a ESM is like that), right?

Comment (guybedford) [reply spec.html:8447]: Yes, the pending state still stands - this is for the case on the 7th slide of https://docs.google.com/presentation/d/1--k5tJmhXJeLlUOmBdHGK3my9VCbZVL0MF8OrTeqS0I/edit?usp=sharing.

File: spec.html (hunk 2 above)
Comment (guybedford) [reply spec.html:22051]: Sure, thanks.

**Review by bmeck — APPROVED** [review]

**Comment by guybedford** [comment]
I've added another commit to try and clarify the note cases here even further. Would value feedback on the wording here before updating the ECMA262 PR.

**Review by littledan — APPROVED** [review]

**Comment by guybedford** [comment]
I've added an additional commit to this at https://github.com/guybedford/ecma262/pull/3/commits/1b409d3eb0248770efd4abe0f5de917efc30c190, which is basically just reducing a minor spec complexity.

To explain:

* If the module hasn't executed yet, GetModuleNamespace calls GetExportedNames which must return *null* to go into the pending state, which then passes handling of the namespace finalization over to the dynamic modules spec.
* If the module has already executed though, GetModuleNamespace should do the finalization rather than the dynamic modules spec, so GetExportedNames in this case only should actually return the valid export names.
* This means that the definition of a "deferred exports" module isn't just that it returns *null* from GetExportedNames anymore, but that it returns *null* in the right cases (during star reexport lookups or for the module's own namespace when the module is unexecuted and detecting this difference based on the *ns* argument we added to GetExportedNames).

One way to make this a bit simpler would just be to always have the dynamic modules spec create the namespaces, and then to always create namespaces for all modules upfront instead of lazily. This is just in spec text though, so implementations could still make namespaces lazily in their own work if they want to.

The spec text then simplifies as provided in https://github.com/guybedford/ecma262/pull/3/commits/1b409d3eb0248770efd4abe0f5de917efc30c190 (no new argument needed for GetExportedNames, and *null* return happens consistently)

The dynamic modules proposal adjustment for this is at https://github.com/nodejs/dynamic-modules/pull/11/files, where namespaces are always finalized at https://github.com/nodejs/dynamic-modules/pull/11/files#diff-3540caefa502006d8a33cb1385720803R176.

Sorry to drag this one out, have been on and off it over the weekend - would value feedback on the above before a final merge!

**Comment by jdalton** [comment]
I'm not a fan of making `export * from "cjs"` throw.
Please see my comment https://github.com/nodejs/dynamic-modules/pull/11#issuecomment-446476928.

**Comment by guybedford** [comment]
As discussed in the modules meeting, I'm going ahead with merging this into the TC39 PR for dynamic modules, while we can continue to discuss alternative approaches.

**Pull request closed without merging**